packages/core/lib/events/DomainEventEmitter.ts

```diff
@@ -151,27 +150,82 @@ export class DomainEventEmitter<SupportedEvents extends CommonEventDefinition[]>
   public onMany<EventTypeName extends EventTypeNames<SupportedEvents[number]>>(
     eventTypeNames: EventTypeName[],
     handler: SingleEventHandler<SupportedEvents, EventTypeName>,
+    isBackgroundHandler = false,
   ) {
     for (const eventTypeName of eventTypeNames) {
-      this.on(eventTypeName, handler)
+      this.on(eventTypeName, handler, isBackgroundHandler)
     }
   }
 
   /**
    * Register handler for all events supported by the emitter
    */
-  public onAny(handler: AnyEventHandler<SupportedEvents>) {
-    this.anyHandlers.push(handler)
+  public onAny(handler: AnyEventHandler<SupportedEvents>, isBackgroundHandler = false) {
+    if (isBackgroundHandler) this.anyHandlers.background.push(handler)
+    else this.anyHandlers.foreground.push(handler)
   }
 
-  private addOnHandler(
-    eventTypeName: EventTypeNames<SupportedEvents[number]>,
-    handler: EventHandler,
-  ) {
-    if (!this.eventHandlerMap[eventTypeName]) {
-      this.eventHandlerMap[eventTypeName] = []
+  private async handleEvent<SupportedEvent extends SupportedEvents[number]>(
+    event: CommonEventDefinitionPublisherSchemaType<SupportedEvent>,
+  ): Promise<void> {
+    const eventHandlers = this.eventHandlerMap[event.type] ?? {
+      foreground: [],
+      background: [],
+    }
+
+    const fgHandlers = [...eventHandlers.foreground, ...this.anyHandlers.foreground]
```

Comment:
it seems a bit wasteful to repopulate this on every event. can we do that after handlers are attached instead?

Reply:
Discussed on Slack, for visibility:
To avoid creating this array with every event, we moved any handler to `eventHandlerMap`

Reply:
Already done :D